cinema_project/base_app/views.py

```diff
@@ -83,41 +26,145 @@ def contact_page(request):
+                                f"<p> Phone: {contact_context['phone']}</p>",
+                   )
         contact_message = ContactMessages(**contact_context)
         contact_message.save()
         return render(request, template_name='contact.html',
                       context=contact_context)
     return render(request, template_name='contact.html')
 
 
-def bookings(request):
-    schedules = Schedule.objects.filter(schedule_time__gte=today()).order_by('schedule_time')
-    print(schedules)
-    return render(request, template_name='bookings.html', context={'bookings': schedules})
-
-
-def activate(request, uidb64, token):
-    User = get_user_model()
-    try:
-        uid = force_str(urlsafe_base64_decode(uidb64))
-        user = User.objects.get(pk=uid)
-    except(TypeError, ValueError, OverflowError, User.DoesNotExist):
-        user = None
-    if user is not None and account_activation_token.check_token(user, token):
-        user.is_active = True
-        user.save()
-        return HttpResponse('Thank you for your email confirmation. Now you can login your account.')
-    else:
-        return HttpResponse('Activation link is invalid!')
-
-
-@ratelimit(key='ip', rate='2/h')
-def fetch_playing_movies(request):
+@ratelimit(key='ip', rate='100/h', block=True)
+def fetch_currently_playing_movies_page(request):
     next_7_days = next_days(7)
     schedules = Schedule.objects.filter(schedule_time__range=[today(), next_7_days])
     paginator = Paginator(schedules, per_page=5)
     page_number = request.GET.get('page')
     page_obj = paginator.get_page(page_number)
     return render(request, template_name="now_playing.html",
                   context={"page_obj": page_obj})
+
+
+def bookings_page(request):
+    cinema_cities = list(set([cinema.city for cinema in Cinema.objects.all()]))
+    cinema_cities.sort()
+    selected_city = cinema_cities[0]
+    if request.method == 'POST':
+        if request.POST.get('selected_city'):
+            selected_city = request.POST['selected_city']
+            return redirect('city_filtered_page', selected_city)
+    return render(request, template_name='bookings.html', context={'cinema_cities': cinema_cities,
+                                                                   'selected_city': selected_city})
+
+
+def city_filtered_page(request, selected_city):
+    schedules_by_city = Schedule.objects.filter(schedule_time__range=[today(), next_days(7)])\
+        .filter(hall__cinema__city__iexact=selected_city)
+    cinema_names = list(set(schedule.hall.cinema.name for schedule in schedules_by_city))
+    cinema_names.sort()
+    selected_cinema = cinema_names[0]
+    if request.method == 'POST':
+        selected_cinema = request.POST['selected_cinema']
+        cinema_names.pop(cinema_names.index(selected_cinema))
+        cinema_names.insert(0, selected_cinema)
+    schedules_by_city_and_cinema = Schedule.objects.filter(hall__cinema__city__iexact=selected_city) \
```

Comment:
This can be written as:
```python
Schedule.objects.filter(
    hall__cinema__city__iexact=selected_city,
    hall__cinema__name__iexact=selected_cinema,
    schedule_time__range=[today(), next_days(7)],
)
```